### PR TITLE
UIROLES-75: show only capabilities/capabilitySets for enabled (versions of) applications,  use chunked requests on capabilities, capability-sets 

### PR DESCRIPTION
**Pull request by aidynoJ**
[Refs UIROLES-75](https://folio-org.atlassian.net/browse/UIROLES-75) - Only display capabilities/capabilitySets for enabled (versions of) applications. Gracefully handle the case where there are so many applications that the query grows too large.
Also it is kind of follow up fix of [UIROLES-60]( https://folio-org.atlassian.net/browse/UIROLES-60) -  the loading issue on create role

### CODE REVIEW

**Review by zburke — COMMENTED** [review]
From the angle of correctness, I think this PR LGTM. I expected the query changes and the introduction of useChunkedCQLFetch, but there's more going on here and it's difficult to follow since the original code isn't well-commented. Tests and documentation are a little too light to approve as-is.

File: src/settings/components/CreateEditRole/CreateEditRoleForm.js [hunk 1]
```diff
@@ -57,10 +57,8 @@ function CreateEditRoleForm({
     onClick={onSubmit}
   ><FormattedMessage id="ui-authorization-roles.crud.saveAndClose" /></Button>;
 
-  const intl = useIntl();
-
   return <form onSubmit={onSubmit} data-testid="create-role-form">
-    <Layer isOpen inRootSet contentLabel={intl.formatMessage({ id: title })}>
+    <Layer isOpen inRootSet contentLabel={<FormattedMessage id={title} />}>
```
[the remark this answers is not in the document]
Comment (aidynoJ) [reply src/settings/components/CreateEditRole/CreateEditRoleForm.js]: Yes, sorry. Accidentally reverted when updated the branch

File: src/settings/components/CreateEditRole/CreateRole.test.js [hunk 2]
```diff
@@ -118,18 +119,6 @@ describe('CreateRole component', () => {
...
-
-    expect(getAllByRole('checkbox')).toHaveLength(3);
-
-    await waitFor(async () => {
-      await userEvent.click(getAllByRole('checkbox')[0]);
-      expect(mockSetSelectedCapabilitiesMap).toHaveBeenCalledWith({ '6e59c367-888a-4561-a3f3-3ca677de437f': true });
-      expect(mockSetSelectedCapabilitiesMap).toHaveBeenCalledTimes(1);
-    });
-  });
-
```
[the remark this answers is not in the document]
Comment (aidynoJ) [reply src/settings/components/CreateEditRole/CreateRole.test.js]: Fixed

File: src/hooks/useApplicationCapabilities.js [hunk 3]
```diff
@@ -15,102 +13,72 @@ import { getCapabilitiesGroupedByTypeAndResource,
  * @returns {Object} An object containing the checkedAppIdsMap and onSubmitSelectApplications function.
  */
 const useApplicationCapabilities = () => {
-  const stripes = useStripes();
-  /* isInitialLoaded is the state that indicates capabilitySets and capabilities data is loaded, since they are depend on each other.
-  Based on this we show spinner on accordion
-  */
```
[the remark this answers is not in the document]
Comment (aidynoJ) [reply src/hooks/useApplicationCapabilities.js:21]: Yes, I agree. I didn't notice how it has been grown. I split it into 2 separate hooks for capabilities, capability-sets. And provided descriptive JS Doc

File: src/hooks/useCapabilities.js [hunk 4]
```diff
@@ -1,21 +1,34 @@
...
-
-import { useNamespace, useOkapiKy } from '@folio/stripes/core';
+import { useChunkedCQLFetch, useNamespace, useStripes } from '@folio/stripes/core';
 
 import { CAPABILITES_LIMIT } from './constants';
 
+// When fetching from a potentially large list of capabilities,
+// single application can include more that 2000 capabilities,
+// make sure to chunk the request to avoid hitting limits.
+
```
[the remark this answers is not in the document]
Comment (aidynoJ) [reply src/hooks/useCapabilities.js]: Got it. The issue with the length of the query, I suggest) Thank you

File: src/settings/components/CreateEditRole/CreateRole.js [hunk 5]
```diff
@@ -38,12 +41,11 @@ const CreateRole = () => {
 
     setDisabledCapabilities({ ...disabledCapabilities, ...capabilitySetsCap });
     setSelectedCapabilitySetsMap({ ...selectedCapabilitySetsMap, [capabilitySetId]: event.target.checked });
-    setSelectedCapabilitiesMap({ ...selectedCapabilitiesMap, ...capabilitySetsCap });
   };
 
-  const isCapabilitySelected = id => !!selectedCapabilitiesMap[id];
   const isCapabilitySetSelected = id => !!selectedCapabilitySetsMap[id];
   const isCapabilityDisabled = id => !!disabledCapabilities[id];
+  const isCapabilitySelected = id => !!selectedCapabilitiesMap[id] || isCapabilityDisabled(id);
```
[the remark this answers is not in the document]
Comment (aidynoJ) [reply src/settings/components/CreateEditRole/CreateRole.js:53]: Added explanation

**Review by zburke — CHANGES_REQUESTED** [review]
Unfortunately, the role-create and role-edit pages no longer populate the "Capability sets" or "Capabilities" accordions for me. I can see the queries in the network inspector, but the UI does not show any data.

Compared to the previous commit, the documentation changes are perfect and helpful and I have resolved all those comments. But now we need to figure out what else changed in the refactor.

**Comment by aidynoJ** [comment]
> Unfortunately, the role-create and role-edit pages no longer populate the "Capability sets" or "Capabilities" accordions for me. I can see the queries in the network inspector, but the UI does not show any data.
> 
> 
> 
> Compared to the previous commit, the documentation changes are perfect and helpful and I have resolved all those comments. But now we need to figure out what else changed in the refactor. 

Interesting, I'll check that out. Thank you

**Review by zburke — APPROVED** [review]
Ohhhhhhh, I understand how this works now, and I apologize for blocking this when it has worked correctly for so long :( The UI here is not intuitive, but that's not your fault. When creating a new Role, in order to show available Capability Sets it is necessary to first select Applications from Capability accordion. This is not intuitive, but it does work.

Since selected Applications determine the available Capability sets and Capabilities, Applications need to be displayed in a hierarchy *above* both, like so:
```
> Application list
  * [ ] App A
  * [*] App B
  > Capability sets
    * [ ] Set A
    * [ ] Set B
  > Capabilities
    * [ ] Capability A
    * [ ] Capability B
    * [ ] Capability C

```

**Pull request merged**